pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go

```diff
@@ -0,0 +1,298 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package kvcoord
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/rpc"
+	"github.com/cockroachdb/cockroach/pkg/util"
+	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/errors"
+)
+
+// rangefeedMuxer is responsible for coordination and management of mux
+// rangefeeds. rangefeedMuxer caches MuxRangeFeed stream per node, and executes
+// each range feed request on an appropriate node.
+type rangefeedMuxer struct {
+	// eventCh receives events from all active muxStreams.
+	eventCh chan *roachpb.MuxRangeFeedEvent
+
+	// Context group controlling execution of MuxRangeFeed calls.
+	g ctxgroup.Group
+
+	// State pertaining to actively executing MuxRangeFeeds.
+	mu struct {
+		syncutil.Mutex
+
+		// terminalErr set when a terminal error occurs.
+		// Subsequent calls to startMuxRangeFeed will return this error.
+		terminalErr error
+
+		// Each call to start new range feed gets a unique ID which is echoed back
+		// by MuxRangeFeed rpc.  This is done as a safety mechanism to make sure
+		// that we always send the event to the correct consumer -- even if the
+		// range feed is terminated and re-established rapidly.
+		nextStreamID int64
+
+		// muxClient contains a nodeID->MuxRangeFeedClient.
+		muxClients map[roachpb.NodeID]*muxClientState
+		// producers maps streamID to the event producer -- data sent back to the
+		// consumer of range feed events.
+		producers map[int64]*channelRangeFeedEventProducer
+	}
+}
+
+type muxClientState struct {
+	client  roachpb.Internal_MuxRangeFeedClient
+	streams util.FastIntSet
+	cancel  context.CancelFunc
+}
+
+func newRangefeedMuxer(g ctxgroup.Group) *rangefeedMuxer {
+	m := &rangefeedMuxer{
+		eventCh: make(chan *roachpb.MuxRangeFeedEvent),
+		g:       g,
+	}
+	m.mu.muxClients = make(map[roachpb.NodeID]*muxClientState)
+	m.mu.producers = make(map[int64]*channelRangeFeedEventProducer)
+
+	m.g.GoCtx(m.demux)
+
+	return m
+}
+
+type eventOrError struct {
+	event *roachpb.RangeFeedEvent
+	err   error
+}
+
+// channelRangeFeedEventProducer is a rangeFeedEventProducer which receives
+// events on input channel, and returns events when Recv is called.
+type channelRangeFeedEventProducer struct {
+	ctx   context.Context
+	errCh chan error // Signalled to propagate terminal error the consumer.
+	input chan eventOrError
```

Comment:
The demuxer synchronously sends on this unbuffered channel without a timeout (in `demux`), isn't that brittle? If any consumer slows down, all other consumers get stuck in the cross-fire. I understand that we don't want to add random buffers everywhere - after all, that's what we're setting out to avoid - but aren't we worried that there will be some reasons for rangefeeds to hang in an extremely opaque way due to this sharing of the demuxer goroutine?

As a first pass, we could have a timeout for sending to a producer's input channel, and if the timeout fires, drop the event & kill that producer. That way, a stuck producer would only momentarily hold up other, more vigilantly consumed producers.

Reply:
>  If any consumer slows down, all other consumers get stuck in the cross-fire.

I think this works as intended.  This is an important part of a pushback mechanism (in changefeeds); We want to slow down ingestion if the consumer cannot consume; (This slow down is also the raison d'être for the existence of this PR 
(since when this slow down happens, data may still be accumulating in the underlying RPC buffers)

Reply:
I was worried about two changefeeds (unrelated sinks) contending, but am now realizing that in `RangeFeedSpans` we make a new demuxer every time, meaning they won't share resources. I agree that this works as advertised then.